Automattic/wp-calypso
# Modularise memberships state

This PR is one of many working on modularising state in Calypso. This one handles memberships.

For more details on state modularisation, see the [modularised state documentation](https://github.com/Automattic/wp-calypso/blob/master/docs/modularized-state.md) and p4TIVU-9lM-p2.

To make memberships state more consistent and reusable, this PR also refactors inline selectors, moving them to dedicated modules. It also removes some unused selectors.

The selectors could probably be rewritten to use optional chaining, but I decided that was out of scope for this PR.

Note: I added all the reviewers that GitHub suggested. Please feel free to ignore the review request or pull in someone else if you're not comfortable reviewing this PR. Thank you!

Fixes #42457.

#### Changes proposed in this Pull Request

* Modularise memberships state
* Extract inline selectors into dedicated selector modules under `state/memberships`
* Remove two unused selectors

#### Testing instructions

* Install [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en).
* Open DevTools and switch to the Redux tab.
* Open `/read` on the live branch or your local copy of this branch.
* Click `State` on the top right to ensure you see all state, and not just the diff for the last action: 
![image](https://user-images.githubusercontent.com/409615/73774406-d3e87d80-477b-11ea-9f59-3e42cc00101f.png)
* Analyse the tree and note that there's no `memberships` key, even if you expand the list of keys. This indicates that the memberships state hasn't been loaded yet.
* Click your avatar on the masterbar, followed by `Manage Purchases` on the sidebar.
* Verify that a `memberships` key is added with the memberships state. This indicates that the memberships state has now been loaded.
* Verify that memberships-related functionality works normally. This indicates that I didn't mess up.
